### PR TITLE
Create K8s pod metric based logs and alerts

### PR DESCRIPTION
**Pull request by malancas**
Closes https://github.com/sigstore/public-good-instance/issues/703

#### Summary

**Summary of metrics and alert policies**

This PR adds two log based metrics for K8s pod errors for Rekor and Fulcio (so four new logs are added in total). It also adds alert policies around each metric.

The metrics simply count the number log entries that contain the specified k8s error message (either "unscheduable" or "Back-off restarting failed container") where `resource.type` is `k8s_pod` and `resource.labels.namespace_name` is either `fulcio-system` or `rekor-system` (depending on the specific metric).

The alert policies on these metrics will fire when logs with the error messages are present for more than ten minutes.

The metrics and alert policies were first created in the GCP UI and converted to Terraform with the terraformer tool. Below are links to the metrics and alert policies in the GCP UI.

- All four user-defined metrics can be found [here](https://console.cloud.google.com/logs/metrics?referrer=search&project=projectsigstore-staging)
- [Fulcio K8s Unscheduable alert policy](https://console.cloud.google.com/monitoring/alerting/policies/17630789102851815676?project=projectsigstore-staging)
- [Fulcio K8s Restart Failing Container alert policy](https://console.cloud.google.com/monitoring/alerting/policies/4640357769451073829?project=projectsigstore-staging)
- [Rekor K8s Unscheduable alert policy](https://console.cloud.google.com/monitoring/alerting/policies/10409889119693812241?project=projectsigstore-staging)
- [Rekor K8s Restart Failing Container alert policy](https://console.cloud.google.com/monitoring/alerting/policies/12865964473412950998?project=projectsigstore-staging)

**Notes on the approach**

https://github.com/sigstore/public-good-instance/issues/703 mentioned investigating whether we could use the GCP Error Reporting service, since it is automatically capturing the K8s pod errors. After some digging, that service does not seem to be supported in the Google Terraform provider and I was unable to configure PagerDuty notifications for it. So I opted to create these logs and alerts manually. These metrics and alerts are almost identical to each other, so it would be good to investigate whether we can reuse some of these metric and alert definitions but for the first pass, I opted to just create separate resource definitions for each metric and alert.

**Questions for the reviewer:**
- Do we want to create these metrics and alerts for the other K8s deployments in the project, like the prober? I just added metrics and alerts to Rekor and Fulcio to start. We can add more in this PR or create a follow up issue.

#### Release Note

None.

#### Documentation

None.

### CODE REVIEW

File: terraform/gcp/modules/monitoring/fulcio/fulcio_alerts.tf [hunk 1]
```diff
@@ -174,3 +174,75 @@ resource "google_monitoring_alert_policy" "ca_service_cert_quota" {
...
+  alert_strategy {
+    auto_close = "604800s"
+  }
+
+  combiner = "OR"
+
+  conditions {
+    condition_threshold {
+      filter     = "metric.name=\"k8s_pod/restarting-failed-container\" resource.type=\"k8s_pod\""
+      duration   = "600s"
```
Comment (malancas) [inline terraform/gcp/modules/monitoring/fulcio/fulcio_alerts.tf:192]: For the first pass, these alerts will fire if the error message is found in logs for at least ten minutes. Let me know if we would like to change this threshold.

File: terraform/gcp/modules/monitoring/rekor/metrics.tf [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+ * limitations under the License.
+ */
+
+# This file contains alerts generic to the Sigstore project
+# Alerts specific to fulcio, rekor or dex should be in the appropriate `modules/monitoring/[service]` directory
+
+resource "google_logging_metric" "rekor_k8s_pod_restart_failing_container" {
+  name   = "rekor_k8s_pod/restarting-failed-container"
+  filter = "resource.labels.namespace_name=\"rekor-system\" resource.type=k8s_pod AND severity>=WARNING \"Back-off restarting failed container\""
+  metric_descriptor {
```
Comment (malancas) [inline terraform/gcp/modules/monitoring/rekor/metrics.tf:23]: These are basic counter metrics, see https://cloud.google.com/logging/docs/logs-based-metrics/counter-metrics for more details.

File: terraform/gcp/modules/monitoring/fulcio/fulcio_alerts.tf [hunk 3]
```diff
@@ -174,3 +174,75 @@ resource "google_monitoring_alert_policy" "ca_service_cert_quota" {
...
+
+  alert_strategy {
+    auto_close = "604800s"
+  }
+
+  combiner = "OR"
+
+  conditions {
+    condition_threshold {
+      filter     = "metric.name=\"k8s_pod/restarting-failed-container\" resource.type=\"k8s_pod\""
```
Comment (priyawadhwa) [inline terraform/gcp/modules/monitoring/fulcio/fulcio_alerts.tf:191]: Do we need to include any fulcio specific details in this filter? Sorry if that's a dumb question I'm not super familiar with the logging alerts

File: terraform/gcp/modules/monitoring/rekor/rekor_alerts.tf [hunk 4]
```diff
@@ -54,3 +54,75 @@ resource "google_monitoring_alert_policy" "rekor_uptime_alerts" {
...
+
+  alert_strategy {
+    auto_close = "604800s"
+  }
+
+  combiner = "OR"
+
+  conditions {
+    condition_threshold {
+      filter     = "metric.name=\"k8s_pod/restarting-failed-container\" resource.type=\"k8s_pod\""
```
Comment (priyawadhwa) [inline terraform/gcp/modules/monitoring/rekor/rekor_alerts.tf:71]: Same question as the above one for fulcio!

File: terraform/gcp/modules/monitoring/fulcio/fulcio_alerts.tf (hunk 3 above)
Comment (malancas) [reply terraform/gcp/modules/monitoring/fulcio/fulcio_alerts.tf]: This filter was incorrect, the metric name portion should have been `fulcio/k8s_pod/restarting-failed-container`. It has been updated to reference this name.

**Review by priyawadhwa — APPROVED** [review]

**Comment by priyawadhwa** [comment]
@malancas i can merge this whenever you think it's ready, just lmk!

**Comment by malancas** [comment]
@priyawadhwa are you fine with merging and provisioning this on Monday on the off the chance these alerts go off over the weekend and need to be adjusted a bit? Just so the channel doesn't get flooded.

**Comment by priyawadhwa** [comment]
@malancas sounds good! i can merge when the tests turn green and we can update the terraform commit in public-good-instance on monday!

**Comment by var-sdk** [comment]
@malancas I ran this today and there were failures during provisioning:

https://github.com/sigstore/public-good-instance/actions/runs/3178209563/jobs/5179470236

**Comment by malancas** [comment]
@var-sdk thanks, I will look into this and get a fix ready. Is the provisioning failure blocking anything that needs to be updated or deployed today?

**Comment by var-sdk** [comment]
Nothing urgent. It can wait until tomorrow.

**Comment by malancas** [comment]
@var-sdk I think this is caused by two different issues so I'll create two PRs for review.

1. The GHA service account is missing an IAM role that allows it to create log based metrics - [solution PR here](https://github.com/sigstore/public-good-instance/pull/858)
2. The alert policy filter needs to be updated because the API Terraform uses to construct the policies is different than the one used by the GCP alert policy creation UI. No error was encountered in the UI while creating these - [solution PR here](https://github.com/sigstore/scaffolding/pull/400)

**Pull request merged**